src/librbd/AioRequest.h

```diff
@@ -134,38 +133,65 @@ namespace librbd {
-     * layering:
+     * layering and the object map:
      *
-     *                           need copyup
-     * LIBRBD_AIO_WRITE_GUARD ---------------> LIBRBD_AIO_WRITE_COPYUP
-     *           |        ^                              |
-     *           v        \------------------------------/
-     *         done
-     *           ^
-     *           |
-     * LIBRBD_AIO_WRITE_FLAT
+     * <start>
+     *  .  |
+     *  .  |
+     *  .  \---> LIBRBD_AIO_WRITE_PRE
+     *  .           |         |
+     *  . . . . . . | . . . . | . . . . . . . . . . . 
+     *      .       |   -or-  |                     .
+     *      .       |         |                     v
+     *      .       |         \----------------> LIBRBD_AIO_WRITE_FLAT . . .
+     *      .       |                                               |      .
+     *      v       v         need copyup                           |      .
+     * LIBRBD_AIO_WRITE_GUARD -----------> LIBRBD_AIO_WRITE_COPYUP  |      .
+     *  .       |   ^                           |                   |      .
+     *  .       |   |                           |                   |      .
+     *  .       |   \---------------------------/                   |      .
+     *  .       |                                                   |      .
+     *  .       \-------------------\           /-------------------/      .
+     *  .                           |           |                          .
+     *  .                       LIBRBD_AIO_WRITE_POST                      .
+     *  .                                |                                 .
+     *  .                                v                                 .
+     *  . . . . . . . . . . . . . . > <finish> < . . . . . . . . . . . . . . 
      *
-     * Writes start in LIBRBD_AIO_WRITE_GUARD or _FLAT, depending on whether
-     * there is a parent or not.
+     * The _PRE_REMOVE/_POST_REMOVE states are skipped if the object map
+     * is disabled.  The write starts in _WRITE_GUARD or _FLAT depending on
+     * whether or not there is a parent overlap.
      */
     enum write_state_d {
       LIBRBD_AIO_WRITE_GUARD,
       LIBRBD_AIO_WRITE_COPYUP,
-      LIBRBD_AIO_WRITE_FLAT
+      LIBRBD_AIO_WRITE_FLAT,
+      LIBRBD_AIO_WRITE_PRE,
+      LIBRBD_AIO_WRITE_POST,
+      LIBRBD_AIO_WRITE_ERROR
     };
 
   protected:
-    virtual void add_copyup_ops() = 0;
-
     write_state_d m_state;
     vector<pair<uint64_t,uint64_t> > m_object_image_extents;
     uint64_t m_parent_overlap;
     librados::ObjectWriteOperation m_write;
-    librados::ObjectWriteOperation m_copyup;
     uint64_t m_snap_seq;
     ceph::bufferlist *m_entire_object;
 
+    virtual void add_write_ops(librados::ObjectWriteOperation *wr) = 0;
+    virtual void guard_write();
+    virtual void pre_object_map_update(uint8_t *new_state) = 0;
+    virtual bool post_object_map_update() {
+      return false;
+    }
+
   private:
+    librados::IoCtx m_io_ctx;
```

Comment:
looks like this ended up not being used (which is fine since we specify snapcontext or snapid for each io)